File: examples/hooks/python/post_tool_log.py
#!/usr/bin/env python3
"""PostToolUse — append JSONL audit log (Python)."""

from __future__ import annotations

import json
import os
import sys
from datetime import datetime, timezone
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent))
from lib.hook_io import allow, read_event  # noqa: E402


def main() -> None:
    event = read_event()
    work_dir = event.get("work_dir") or os.getcwd()
    log_path = Path(
        os.environ.get("SOLCODE_HOOK_LOG")
        or Path(work_dir) / ".solcode" / "hooks-tool.jsonl"
    )
    result = event.get("tool_result") or {}
    text = ""
    if isinstance(result, dict) and isinstance(result.get("text"), str):
        text = result["text"]

    line = {
        "ts": datetime.now(timezone.utc).isoformat(),
        "event": event.get("event"),
        "tool": event.get("tool_name"),
        "session_id": event.get("session_id"),
        "is_error": bool(result.get("is_error")) if isinstance(result, dict) else False,
        "text_chars": len(text),
        "text_preview": text[:200],
        "input": event.get("tool_input"),
    }
    try:
        log_path.parent.mkdir(parents=True, exist_ok=True)
        with log_path.open("a", encoding="utf-8") as f:
            f.write(json.dumps(line, ensure_ascii=False) + "\n")
    except OSError as err:
        print(f"post_tool_log write failed: {err}", file=sys.stderr)
    allow()


if __name__ == "__main__":
    main()

File: examples/hooks/post-tool-log.mjs
#!/usr/bin/env node
/**
 * PostToolUse — append a JSONL audit log of tool calls
 *
 * Does not modify results (decision: allow). Useful next to the builtin
 * compress_tool_result hook.
 *
 * {
 *   "hooks": {
 *     "events": {
 *       "PostToolUse": [{
 *         "matcher": "*",
 *         "hooks": [{
 *           "type": "command",
 *           "command": "node examples/hooks/post-tool-log.mjs",
 *           "timeout_ms": 3000,
 *           "fail_mode": "open"
 *         }]
 *       }]
 *     }
 *   }
 * }
 *
 * Log file: $SOLCODE_HOOK_LOG or <work_dir>/.solcode/hooks-tool.jsonl
 */

import { appendFileSync, mkdirSync } from "node:fs";
import { dirname, join } from "node:path";
import { readEvent, allow } from "./lib/read-event.mjs";

const event = await readEvent();
const workDir = event.work_dir || process.cwd();
const logPath =
  process.env.SOLCODE_HOOK_LOG ||
  join(workDir, ".solcode", "hooks-tool.jsonl");

const result = event.tool_result || {};
const text =
  typeof result === "object" && result && typeof result.text === "string"
    ? result.text
    : "";

const line = {
  ts: new Date().toISOString(),
  event: event.event,
  tool: event.tool_name,
  session_id: event.session_id,
  is_error: Boolean(result.is_error),
  text_chars: text.length,
  text_preview: text.slice(0, 200),
  input: event.tool_input,
};

try {
  mkdirSync(dirname(logPath), { recursive: true });
  appendFileSync(logPath, JSON.stringify(line) + "\n", "utf8");
} catch (err) {
  // fail_mode open: still allow the tool result through
  console.error("post-tool-log write failed:", err.message);
}

allow();
